dali/pipeline/operators/util/external_source.cc

```diff
@@ -20,12 +20,20 @@ template<>
 void ExternalSource<CPUBackend>::RunImpl(SampleWorkspace *ws, const int idx) {
   // Wrap the output tensor around our data
   auto output = ws->Output<CPUBackend>(idx);
+  cudaStream_t stream = ws->has_stream() ? ws->stream() : 0;
   if (data_in_tl_) {
-    output->ShareData(&tl_data_, ws->data_idx());
+    output->Copy(tl_data_, ws->data_idx(), stream);
   } else {
     DALI_ENFORCE_VALID_INDEX(ws->data_idx(), t_data_.size());
     auto &data = t_data_[ws->data_idx()];
-    output->ShareData(&data);
+    output->Copy(data, stream);
+  }
+  ++samples_processed_;
+
+  if (samples_processed_.load() == batch_size_) {
+    samples_processed_ = 0;
```

Comment:
Is there still some possibility of a race condition here?  If the thread is interrupted before line 34 then we could increment twice, set to zero twice, and possibly blow past == batch_size_.  No?

Reply:
` if (samples_processed_.load() + 1 == batch_size_) {
    samples_processed_ = 0;

Reply:
Normally if one of the threads is interrupted, all the other threads are interrupted too.
But I agree that for thread safety, I should switch `std::atomic` with `std::mutex` to change to:
```
{
  std::unique_lock<std::mutex> lock(sp_m_);
  if (samples_processed_ + 1 == batch_size_) {
    samples_processed_ = 0;
    [...]
  } else {
    ++samples_processed_;
  }
}
```

Reply:
```
  if (++samples_processed_ >= batch_size_) {
    samples_processed_ = 0;
    [...]
  }
```
Maybe?

Reply:
@cliffwoolley - there is a pitfall when one thread enters the body of it, is interrupted and other ones keep incrementing samples_processed_. When it is resumed it sets samples_processed_ to 0 while it should be samples_processed_ -= batch_size_ as I understand. Still putting just samples_processed_ -= batch_size_ in the body is wrong as multiple threads can enter it and each of them will subtract batch_size_.
So mutex seems to be working solution or I'm missing something.

Reply:
Sorry yes - I was unclear.  The mutex is good.

My most recent comment was only on the thing inside the mutexed region, as in it could just be if and not if/else.

Reply:
Sure, then I total agree.

Reply:
Actually, one more question about this -- how does the `idx` parameter related to `samples_processed_`?  Can multiple threads end up getting assigned the same `idx` and then effectively skipping over some value of `samples_processed_` if the two threads are running in more-or-less lock-step?